### ISSUE TITLE
the same data, the same code , different thread number, different result, why ?

### ISSUE THREAD

**Issue by yale8848** [facebookresearch/fastText#315]
the same data, the same code , different thread number, different result, why ?

CentOS7  x64:

./fasttext supervised -input train.dat -output model -epoch 25 -loss softmax -ws 5 -thread 10
./fasttext test model.bin test.dat
P@1        0.906
R@1	0.906

./fasttext supervised -input train.dat -output model -epoch 25 -loss softmax -ws 5 -thread 20
./fasttext test model.bin test.dat
P@1    	0.939
R@1	0.939

**Comment by pommedeterresautee** [comment]
you don t need to change the number of thread to get different results. Init of vectors is a random process.

**Comment by yale8848** [comment]
thanks @pommedeterresautee  , but i don t understand why design to random , and it will influence of the test results

**Comment by MarcSzafraniec** [comment]
I have the same problem, very wide variations on classification results (up to 0.2 ROC AUC) with the exact same parameters. My dataset has approx 450K texts. Someone knows how to get more consistent results?

**Comment by cpuhrsch** [comment]
Helo @yale8848,

Thank you for your post. Unfortunately we cannot help you with this issue as we don't have access to the data. For me to reproduce this behavior I will either need access to this data or find it reproduced on one of our test sets. Further, we have not observed this behavior as part of our integration tests. In general, it is indeed true that we randomly initialize the parameters and also that the number of threads should have a small influence (in either direction) on the test performance. This stems from the fact that we updated the parameters without regard for synchronization to improve performance. See the paper on [Hogwild](https://people.eecs.berkeley.edu/~brecht/papers/hogwildTR.pdf) for details.

Thanks,
Christian

**Comment by yale8848** [comment]
Dear @cpuhrsch, 

Get it, i will make some small data to test this, if this problem still appear , i will show the data. 

Thank you very much,
Yale

**Comment by suresh-soundararajan** [comment]
Hi Team,

I am also facing with the same issue. I am tried to achieve binary classification use case. Whenever I build the model with the same set of training data, the loss value is getting huge difference. Each time the results are drastically deviating.

with the -lr 0.5 -epoch 50 -loss ns

Am I missing any parameter.

Your help is much appreciated.

Thanks,
SureshKumar S

**Comment by simon19891101** [comment]
Facing the same issue for the supervised model. This issue happens a lot when loss='hs'. 

Params are: lr=0.05, wordNgrams=3, epoch=10, verbose=2, dim=300, loss='hs'.

Reall@1 for 5 training runs can be like: 0.68, 0.72, 0.75, 0.69, 0.76.

Any help welcomed! Cheers.
